File: app/src/main/java/com/androidizate/clase8/MainActivity.java
package com.androidizate.clase8;

import android.content.Context;
import android.content.DialogInterface;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.TouchDelegate;

import com.androidizate.clase8.adapters.UserAdapter;
import com.androidizate.clase8.dao.User;

import java.util.List;

public class MainActivity extends AppCompatActivity implements MainView {

    RecyclerView recyclerView;
    MainPresenter presenter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        recyclerView = (RecyclerView) findViewById(R.id.recycler);

        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(linearLayoutManager);

        presenter = new MainPresenter();
    }

    @Override
    protected void onResume() {
        super.onResume();
        presenter.setView(this);
        presenter.onResume();
    }

    @Override
    protected void onPause() {
        super.onPause();
        presenter.setView(null);
    }

    @Override
    public void createAlert(String message) {
        new AlertDialog.Builder(this)
                .setMessage(String.format(getString(R.string.error), message))
                .setTitle(getString(R.string.error))
                .setPositiveButton(R.string.ok, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        MainActivity.this.finish();
                    }
                }).show();
    }

    @Override
    public void loadData(List<User> users) {
        recyclerView.setAdapter(new UserAdapter(users));
    }

    @Override
    public void clearData() {
        // NO OP
    }

    @Override
    public boolean isNetworkConnected() {
        ConnectivityManager connectivityManager = (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo networkInfo = connectivityManager.getActiveNetworkInfo();
        return networkInfo != null && networkInfo.isConnected();
    }

    @Override
    public String getNoConnectionErrorString() {
        return getString(R.string.not_connected_error);
    }
}

File: app/src/main/java/com/androidizate/clase8/MainActivity.kt
package com.androidizate.clase8

import android.content.Context
import android.net.ConnectivityManager
import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.isVisible
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import com.androidizate.clase8.adapters.PostAdapter
import com.androidizate.clase8.adapters.UserAdapter
import com.androidizate.clase8.repositories.datasources.local.AppDatabase
import com.androidizate.clase8.repositories.datasources.remote.RestApiClient
import com.androidizate.clase8.repositories.datasources.remote.dtos.User
import com.androidizate.clase8.repositories.datasources.remote.dtos.toDbEntity
import com.androidizate.clase8.repositories.datasources.remote.dtos.toUser
import com.androidizate.clase8.repositories.datasources.remote.dtos.toPost
import com.androidizate.clase8.repositories.datasources.remote.dtos.Post
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.coroutines.*
import java.lang.String.*

class MainActivity : AppCompatActivity() {
    private val restApiClient = RestApiClient()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val linearLayoutManager = LinearLayoutManager(this)
        recycler.layoutManager = linearLayoutManager
        if (!isNetworkConnected()) {
            createAlert(getString(R.string.not_connected_error))
            getLocalData()
        } else {
            downloadInfo()
        }
    }

    private fun getLocalData() = lifecycleScope.launch {
        val posts =
            AppDatabase.getInstance(applicationContext).postDao().getAll().map { postEntity ->
                postEntity.toPost()
            }
        recycler.adapter = PostAdapter(posts)
        progressBar.isVisible = false
    }

    private fun createAlert(stringResource: String?) {
        AlertDialog.Builder(this)
            .setMessage(format(getString(R.string.error), stringResource))
            .setTitle(getString(R.string.error))
            .setPositiveButton(R.string.ok) { dialog, i -> dialog.dismiss() }
            .show()
    }

    private fun isNetworkConnected(): Boolean {
        val connectivityManager =
            getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val networkInfo = connectivityManager.activeNetworkInfo
        return networkInfo != null && networkInfo.isConnected
    }

    private fun downloadInfo() = lifecycleScope.launch(Dispatchers.Main) {
        progressBar.isVisible = true
        var posts = emptyList<Post>()
        try {
            posts = withContext(Dispatchers.IO) {
                restApiClient.getAllPosts()
            }
            posts.forEach {
                AppDatabase.getInstance(applicationContext).postDao().insertAll(it.toDbEntity())
            }
        } catch (exception: Exception) {
            Log.e("MainActivity", exception.message.toString())
        }
        recycler.adapter = PostAdapter(posts)
        progressBar.isVisible = false
    }
}
